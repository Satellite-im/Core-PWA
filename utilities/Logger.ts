export enum LogLevel {
  PROD = 'PROD',
  DEV = 'DEV',
}

/**
 * @class
 * @description This method should be used for ALL logging you plan to leave in the application.
 * DO NOT use normal logging, the use of this class will make it easier to enable / disable different log levels.
 */
export default class Logger {
  debug: boolean

  constructor(debug: boolean) {
    this.debug = debug
  }

  /**
   * @method _log
   * @description Internal logging method used to abstract and format logging data.
   * @param tag This is usually the file or feature that is logging
   * @param desc Short description of the log
   * @param data Optional data to log
   * @param level to log to, anything below prod will not be shown unless debug is enabled in the config.
   * @returns Blob stripped of EXIF data
   */
  private _log(tag: string, desc: string, data: object, level: LogLevel) {
    const hasData: boolean = Object.keys(data).length >= 1
    if (level === LogLevel.PROD || this.debug) {
      console[hasData ? 'groupCollapsed' : 'log'](
        `%c${tag}%c${desc}`,
        'color:white; background: #34495e; border-radius: 2px 0 0 2px; padding: 0.1rem 0.5rem; border-right: none;',
        'color:white; background: #3498db; border-radius: 0 2px 2px 0; padding: 0.1rem 0.5rem; border-left: none;',
      )
      if (hasData) console.log(data)
      console.groupEnd()
    }
  }

  /**
   * @method _log
   * @description Log data in a formatted manner, programatically.
   * @param tag This is usually the file or feature that is logging
   * @param desc Short description of the log
   * @param data Optional data to log
   * @param level to log to, anything below prod will not be shown unless debug is enabled in the config.
   * @returns Blob stripped of EXIF data
   */
  log(
    tag: string,
    desc: string,
    data: object = {},
    level: LogLevel = LogLevel.PROD,
  ) {
    this._log(tag, desc, data, level)
  }
}
