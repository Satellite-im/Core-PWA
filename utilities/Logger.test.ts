import Logger from './Logger'

describe('Logger', () => {
  let instance: Logger

  beforeEach(() => {
    instance = new Logger()
  })

  it('instance should be an instanceof Logger', () => {
    expect(instance instanceof Logger).toBeTruthy()
  })

  describe('Spy Logger Methods', () => {
    it('should log [default] correctly', () => {
      const loggerSpy = jest.spyOn(Logger.prototype as any, 'log')
      const loggerInstance = new Logger()

      loggerInstance.log('Logger', 'Description', {})

      expect(loggerSpy).toHaveBeenCalled()
      expect(loggerSpy).toHaveBeenCalledWith('Logger', 'Description', {})
    })

    it('should log info correctly', () => {
      const loggerSpy = jest.spyOn(Logger.prototype as any, 'info')
      const loggerInstance = new Logger()

      loggerInstance.info('Logger', 'Description', {})

      expect(loggerSpy).toHaveBeenCalled()
      expect(loggerSpy).toHaveBeenCalledWith('Logger', 'Description', {})
    })

    it('should log warn correctly', () => {
      const loggerSpy = jest.spyOn(Logger.prototype as any, 'warn')
      const loggerInstance = new Logger()

      loggerInstance.warn('Logger', 'Description', {})

      expect(loggerSpy).toHaveBeenCalled()
      expect(loggerSpy).toHaveBeenCalledWith('Logger', 'Description', {})
    })

    it('should log error correctly', () => {
      const loggerSpy = jest.spyOn(Logger.prototype as any, 'error')
      const loggerInstance = new Logger()

      loggerInstance.error('Logger', 'Description', {})

      expect(loggerSpy).toHaveBeenCalled()
      expect(loggerSpy).toHaveBeenCalledWith('Logger', 'Description', {})
    })

    it('should log debug correctly', () => {
      const loggerSpy = jest.spyOn(Logger.prototype as any, 'debug')
      const loggerInstance = new Logger()

      loggerInstance.debug('Logger', 'Description', {})

      expect(loggerSpy).toHaveBeenCalled()
      expect(loggerSpy).toHaveBeenCalledWith('Logger', 'Description', {})
    })

    it('should log prod correctly', () => {
      const loggerSpy = jest.spyOn(Logger.prototype as any, 'prod')
      const loggerInstance = new Logger()

      loggerInstance.prod('Logger', 'Description', {})

      expect(loggerSpy).toHaveBeenCalled()
      expect(loggerSpy).toHaveBeenCalledWith('Logger', 'Description', {})
    })

    it('should log [default] correctly without data parameter', () => {
      const loggerSpy = jest.spyOn(Logger.prototype as any, 'log')
      const loggerInstance = new Logger()

      loggerInstance.log('Logger', 'Description')

      expect(loggerSpy).toHaveBeenCalled()
      expect(loggerSpy).toHaveBeenCalledWith('Logger', 'Description')
    })

    it('should log info correctly without data parameter', () => {
      const loggerSpy = jest.spyOn(Logger.prototype as any, 'info')
      const loggerInstance = new Logger()

      loggerInstance.info('Logger', 'Description')

      expect(loggerSpy).toHaveBeenCalled()
      expect(loggerSpy).toHaveBeenCalledWith('Logger', 'Description')
    })

    it('should log warn correctly without data parameter', () => {
      const loggerSpy = jest.spyOn(Logger.prototype as any, 'warn')
      const loggerInstance = new Logger()

      loggerInstance.warn('Logger', 'Description')

      expect(loggerSpy).toHaveBeenCalled()
      expect(loggerSpy).toHaveBeenCalledWith('Logger', 'Description')
    })

    it('should log error correctly without data parameter', () => {
      const loggerSpy = jest.spyOn(Logger.prototype as any, 'error')
      const loggerInstance = new Logger()

      loggerInstance.error('Logger', 'Description')

      expect(loggerSpy).toHaveBeenCalled()
      expect(loggerSpy).toHaveBeenCalledWith('Logger', 'Description')
    })

    it('should log debug correctly without data parameter', () => {
      const loggerSpy = jest.spyOn(Logger.prototype as any, 'debug')
      const loggerInstance = new Logger()

      loggerInstance.debug('Logger', 'Description')

      expect(loggerSpy).toHaveBeenCalled()
      expect(loggerSpy).toHaveBeenCalledWith('Logger', 'Description')
    })

    it('should log prod correctly without data parameter', () => {
      const loggerSpy = jest.spyOn(Logger.prototype as any, 'prod')
      const loggerInstance = new Logger()

      loggerInstance.prod('Logger', 'Description')

      expect(loggerSpy).toHaveBeenCalled()
      expect(loggerSpy).toHaveBeenCalledWith('Logger', 'Description')
    })
  })
})
