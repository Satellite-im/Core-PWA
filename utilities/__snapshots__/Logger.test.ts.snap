// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logger.default._log 0 1`] = `undefined`;

exports[`Logger.default._log 1 1`] = `undefined`;

exports[`Logger.default._log 2 1`] = `undefined`;

exports[`Logger.default._log 3 1`] = `undefined`;

exports[`Logger.default._log 4 1`] = `undefined`;

exports[`Logger.default._log 5 1`] = `undefined`;

exports[`Logger.default.log 0 1`] = `undefined`;

exports[`Logger.default.log 1 1`] = `undefined`;

exports[`Logger.default.log 2 1`] = `undefined`;

exports[`Logger.default.log 3 1`] = `undefined`;

exports[`Logger.default.log 4 1`] = `undefined`;

exports[`Logger.default.log 5 1`] = `undefined`;
